File: vuln-sandbox/go/cwe-89/sql-injection-safe.go
// SAFE: go-sql-string-concat — parameterized query prevents SQL injection
// Rule: go-sql-string-concat | CWE-89 | Expected: TrueNegative

package main

import (
	"database/sql"
	"fmt"
	"net/http"

	_ "github.com/lib/pq"
)

var db *sql.DB

func getUserHandler(w http.ResponseWriter, r *http.Request) {
	userId := r.URL.Query().Get("id")

	// SAFE: user input passed as a parameterized query argument, never concatenated
	rows, err := db.Query("SELECT * FROM users WHERE id = $1", userId)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer rows.Close()
	fmt.Fprintf(w, "query executed safely")
}

func main() {
	var err error
	db, err = sql.Open("postgres", "postgres://localhost/app?sslmode=disable")
	if err != nil {
		panic(err)
	}
	http.HandleFunc("/user", getUserHandler)
	http.ListenAndServe(":8080", nil)
}

File: vuln-sandbox/rust/cwe-89/sql_injection_safe.rs
// SAFE: rust-sqlx-format-query — sqlx::query! macro with parameterized binding
// Rule: rust-sqlx-format-query | CWE-89 | Expected: TrueNegative

use sqlx::PgPool;

async fn get_user(pool: &PgPool, user_id: i64) -> Result<String, sqlx::Error> {
    // SAFE: sqlx::query! macro uses compile-time checked parameterized binding;
    // user input is never interpolated into the SQL string via format!
    let row = sqlx::query!(
        "SELECT username FROM users WHERE id = $1",
        user_id
    )
    .fetch_one(pool)
    .await?;

    Ok(row.username)
}

#[tokio::main]
async fn main() {
    let pool = PgPool::connect("postgres://localhost/app").await.unwrap();
    match get_user(&pool, 1).await {
        Ok(name) => println!("user: {}", name),
        Err(e) => eprintln!("error: {}", e),
    }
}
